### PR TITLE
Trie

### PR DESCRIPTION
**Pull request by drmingdrmer**
# Description

Breaking changes:

-   Array: Use int32 for Cnt and Offsets instead of uint32
-   Remove array.Marshal, use proto.Marshal
-   Use marshal instead of array.Converter
-   Simplify NewXX names, e.g.: NewArrayU16 to NewU16
-   Remove unused version test
-   Remove GetVersion
-   Remove Converters

New features:

-   ArrayBase add method GetTo and benchmark

Add Tests:

-   InitIndex
-   Test of array elements with different size
-   Add full test to ArrayBase
-   Test inter-marshal between array.U32 and array.Array

Refectoring:

-   Move benchmark supporting utils into separate package
-   ArrayBase use marshaler
-   Add ArrayBase.Get as general Get method
-   Benchmark use package marshal instead of array.Converter
-   SlimTrie uses prefefined array U16 U32 as its field type: Children, Steps and Leave
-   Fix lint issues
-   Update search benchmark chart

## Type of change

- **Bug fix**         
- **New feature**     
- **Breaking change** 
- **Refactoring**
- **Document changes**
- **Test changes**

# Checklist:

- [x] **Style**:       My code follows the **style guidelines** of this project
- [x] **Self-review**: I have performed a **self-review** of my own code
- [x] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [x] **Doc**:         I have made corresponding changes to the **documentation**
- [x] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes
- [ ] **Dep**:         Any **dependent** changes have been merged and published in downstream modules

### CODE REVIEW

File: array/errors.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package array
+
+import (
+	"github.com/openacid/errors"
+)
+
+// ErrIndexNotAscending means indexes to initialize a Array must be in
```
Comment (liubaohai) [inline array/errors.go]: `an Array `

File: array/errors.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// ascending order.
+var ErrIndexNotAscending = errors.New("index must be an ascending ordered slice")
+
+// ErrIndexLen is returned if number of indexes does not equal the number of
+// datas, when initializing a Array.
```
Comment (liubaohai) [inline array/errors.go]: `an Array`

File: array/array.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"reflect"
+
+	"github.com/openacid/slim/marshal"
+)
+
+// Array is a space efficient array implementation.
+//
+// Unlike a normal array, it does not allocate space for a element that there is
```
Comment (wenbobuaa) [inline array/array.go]:
```suggestion
// Unlike a normal array, it does not allocate space for an element that there is
```

File: array/array_base.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+const (
+	// bmWidth defines how many bits for a bitmap word
+	bmWidth = int32(64)
+	bmMask  = int32(63)
+)
+
+// bmBit calculates bitamp word index and the bit index in the word.
+func bmBit(idx int32) (int32, int32) {
+	c := idx >> uint32(6) // == idx / bmWidth
+	r := idx & int32(63)  // == idx % bmWidth
```
Comment (wenbobuaa) [inline array/array_base.go]:
```suggestion
	r := idx & bmMask  // == idx % bmWidth
```

File: array/array.go (hunk 3 above)
Comment (wenbobuaa) [inline array/array.go]:
```suggestion
// Unlike a normal array, it does not allocate space for an element that there is
```

File: array/array_base.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+const (
+	// bmWidth defines how many bits for a bitmap word
+	bmWidth = int32(64)
+	bmMask  = int32(63)
+)
+
+// bmBit calculates bitamp word index and the bit index in the word.
+func bmBit(idx int32) (int32, int32) {
+	c := idx >> uint32(6) // == idx / bmWidth
```
Comment (wenbobuaa) [inline array/array_base.go]: 这里的6也需要个定义不

File: array/array_base.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+	if ok {
+		_, v := a.EltMarshaler.Unmarshal(bs)
+		return v, true
+	}
+
+	return nil, false
+}
+
+// GetBytes is similar to Get but does not return the byte slice instead of
+// unmarshaled data.
```
Comment (wenbobuaa) [inline array/array_base.go]: 这一句文档是不是写错了，没读明白..

File: array/array_base.go [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+	_, err = a.InitElts(elts, marshaler)
+	if err != nil {
+		return errors.Wrapf(err, "failure Init Array")
+	}
+
+	return nil
+}
+
+func (a *ArrayBase) InitElts(elts interface{}, marshaler marshal.Marshaler) (int, error) {
```
Comment (wenbobuaa) [inline array/array_base.go]: 这里 和后面的 Get 怎么没加文档

File: array/array_base.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	if a.EltMarshaler == nil {
+		var err error
+		marshaler, err = marshal.NewTypeMarshalerEndian(rElts.Index(0).Interface(), endian)
+		if err != nil {
+			// TODO wrap
+			return err
+		}
+	} else {
+		marshaler = a.EltMarshaler
+	}
```
Comment (pengsven) [inline array/array_base.go]: 125~136提前到一个函数里如何？，因为array.go也有同样的代码块

File: array/array.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	a := &Array{}
+	a.EltMarshaler = m
+	return a, nil
+}
+
+// New creates an array from indexes and elts.
+// Length of indexes and length of elts must be the same.
+// elts must be a slice of value of fixed size.
```
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply array/array.go]: https://translate.google.com/#view=home&op=translate&sl=zh-CN&tl=en&text=%E4%BB%96%E4%BB%AC%E5%BF%85%E9%A1%BB%E6%98%AF%E5%A4%A7%E5%B0%8F%E5%9B%BA%E5%AE%9A%E7%9A%84%E5%85%83%E7%B4%A0%E4%B8%80%E4%B8%AA%E6%95%B0%E7%BB%84

File: array/array_base.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+	"encoding/binary"
+	"reflect"
+
+	"github.com/openacid/errors"
+	"github.com/openacid/slim/bits"
+	"github.com/openacid/slim/marshal"
+	"github.com/openacid/slim/prototype"
+)
+
+// endian is default endian for array
```
[the remark this answers is not in the document]
Comment (drmingdrmer) [reply array/array_base.go]: 额.这个是描述下面那个变量`endian`的. 必须以`endian`开头...go 规定的.

File: array/array_base.go (hunk 8 above)
Comment (drmingdrmer) [reply array/array_base.go]: 有类似的.不太一样...能提取的部分很少...

**Pull request merged**